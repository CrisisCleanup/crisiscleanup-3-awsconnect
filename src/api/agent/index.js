/**
 * index.js
 * Agent Module
 */

import * as Agent from './legacy';

export default Agent;
