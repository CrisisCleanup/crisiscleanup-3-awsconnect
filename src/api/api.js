/**
 * api.js
 * Base Api Module
 */

import { Dynamo } from '../utils';

export default class ApiModel {
  constructor({ loggerName = '', dbTable } = {}) {
    this.loggerName = loggerName;
    if (dbTable) {
      this.db = Dynamo.DynamoClient(dbTable);
    }
  }

  log(message, ...params) {
    console.log(`${this.loggerName} `, message, ...params);
    return this;
  }
}
