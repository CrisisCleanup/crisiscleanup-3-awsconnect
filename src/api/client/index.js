/**
 * index.js
 * Client Module
 */

import * as Client from './client';

export default Client;
