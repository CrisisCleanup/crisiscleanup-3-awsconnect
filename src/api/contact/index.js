/**
 * index.js
 * Contact Module
 */

import * as Contact from './contact';

export default Contact;
