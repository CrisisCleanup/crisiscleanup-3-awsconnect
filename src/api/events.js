/**
 * events.js
 * Events Module
 */

import axios from 'axios';

export const EVENT_OBJECTS = Object.freeze({
  AGENT: 'agent',
  INBOUND: 'inbound-call',
  OUTBOUND: 'outbound-call',
});

export const EVENT_ACTIONS = Object.freeze({
  CREATE: 'create',
  UPDATE: 'update',
});

export class Event {
  constructor({ itemId = null } = {}) {
    this.eventItem = {
      object_id: itemId,
      attr: { __source: 'awsconnect' },
    };
    this.eventDef = {
      action: null,
      object: null,
      joined: null,
    };
    this.logObjs = [];
    Object.keys(EVENT_ACTIONS).forEach((k) => {
      this[k.toLowerCase()] = () => {
        const action = EVENT_ACTIONS[k];
        this.log(`setting event action: ${action}`);
        this.eventDef.action = action;
        return this;
      };
    });
  }

  get eventKey() {
    const { object, action, joined } = this.eventDef;
    if (action === 'join' && joined) {
      // eslint-disable-next-line no-underscore-dangle
      this.eventItem.attr.__joined_id = joined.recent;
      return `${action}_${object}_to_${joined.eventDef.object}`;
    }
    return `${action}_${object}`;
  }

  get recent() {
    if (!this.logObjs.length) {
      return null;
    }
    this.log('joining recent log object:', this.logObjs);
    return this.logObjs[this.logObjs.length - 1].id;
  }

  log(message) {
    const title = `event[${this.eventItem.object_id}|${this.eventKey}]`;
    console.log(`${title} ${message}`);
  }

  object(value) {
    console.log(`setting current object: ${value}`);
    this.eventDef.object = value;
    return this;
  }

  join(event) {
    console.log(`joining event to ${event}`);
    this.eventDef.action = 'join';
    this.eventDef.joined = event;
    return this;
  }

  async save(attrs = {}) {
    this.eventItem.event_key = this.eventKey;
    this.eventItem.attr = { ...this.eventItem.attr, ...attrs };
    const resp = await axios.post('/event_logs', this.eventItem);
    this.logObjs.push(resp.data);
    return this;
  }
}
