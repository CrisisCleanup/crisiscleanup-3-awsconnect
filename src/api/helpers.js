/**
 * helpers.js
 * Api Helpers
 */

import axios from 'axios';

const LANGUAGE = {
  en_US: 2,
  es_MX: 7,
};

export const getLanguageId = async (subtag) => {
  if (subtag.includes('-')) {
    return getLanguageId(subtag.replace('-', '_'));
  }
  if (Object.keys(LANGUAGE).includes(subtag)) {
    return LANGUAGE[subtag];
  }
  return 2;
};
