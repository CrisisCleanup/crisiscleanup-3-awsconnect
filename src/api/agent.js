/* eslint-disable camelcase */
/**
 * agent.js
 * Agent Api Module
 */

import { Dynamo } from '../utils';
import { Event, EVENT_OBJECTS } from './events';

const TABLE = Dynamo.TABLES.AGENTS;

export const AGENT_STATES = Object.freeze({
  PENDING_CALL: 'PendingBusy',
  AGENT_CALLING: 'CallingCustomer',
  AGENT_PENDING: 'pending',
  ROUTABLE: 'routable',
  NOT_ROUTABLE: 'not_routable',
  BUSY: 'Busy',
  ON_CALL: 'Busy',
  OFFLINE: 'offline',
  DISCONNECTED: 'ended',
  PAUSED: 'AfterCallWork',
});

export const AGENT_STATE_TYPES = [
  AGENT_STATES.ROUTABLE,
  AGENT_STATES.NOT_ROUTABLE,
];

const ROUTABLE_STATES = [AGENT_STATES.ROUTABLE];

const NOT_ROUTABLE_STATES = [
  AGENT_STATES.ON_CALL,
  AGENT_STATES.NOT_ROUTABLE,
  AGENT_STATES.OFFLINE,
  AGENT_STATES.AGENT_PENDING,
  AGENT_STATES.PAUSED,
  AGENT_STATES.PENDING_CALL,
  AGENT_STATES.AGENT_CALLING,
];

const INROUTE_STATES = [AGENT_STATES.PENDING_CALL, AGENT_STATES.AGENT_CALLING];

export const AGENT_STATE_GROUPS = Object.freeze({
  [AGENT_STATES.ROUTABLE]: ROUTABLE_STATES,
  [AGENT_STATES.NOT_ROUTABLE]: NOT_ROUTABLE_STATES,
});

export const isRoutable = (state) =>
  ROUTABLE_STATES.includes(state.split('#')[2] || state);
export const isInRoute = (state) =>
  INROUTE_STATES.includes(state.split('#')[2] || state);

export const getStateDef = (state) => {
  if (!state || state === null || typeof state !== 'string') {
    console.log(`state ${state} is not a supported type!`);
    return getStateDef(AGENT_STATES.OFFLINE);
  }
  if (state.includes('#')) {
    console.log(`state [${state}] already contains type, returning`);
    return state.split('#');
  }
  const stateType = Object.keys(AGENT_STATE_GROUPS).filter((key) =>
    AGENT_STATE_GROUPS[key].includes(state) ? key : false,
  );
  if (stateType.length) {
    console.log(`found state type: ${stateType} for state: ${state}`);
    const isOnline = state === 'offline' ? 'offline' : 'online';
    return [isOnline, stateType[0], state];
  }
  console.log(`Unknown state type for state: ${state}`);
  return ['offline', '', state];
};

export const AGENT_ATTRS = Object.freeze({
  STATE: 'state',
  ENTERED: 'entered_timestamp',
  LAST_CONTACT: 'last_contact_id',
  CURRENT_CONTACT: 'current_contact_id',
  CONNECTION: 'connection_id',
  ACTIVE: 'active',
  STATE_TTL: 'state_ttl',
});

export const KeyMap = ({
  mapName = 'Key',
  partitionKey = TABLE.hash,
  agentId,
  attributes = {},
}) => ({
  [mapName]: {
    [partitionKey]: {
      S: agentId,
    },
    ...attributes,
  },
});

export const get = async ({ agentId, attributes }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const projExp = attributes || Object.values(AGENT_ATTRS);
  const params = {
    ...KeyMap({
      agentId,
    }),
    ProjectionExpression: projExp.join(','),
  };
  if (projExp.includes('state')) {
    // state is a dynamodb reserved key word
    params.ExpressionAttributeNames = {
      '#A': 'state',
    };
    const filteredProjExp = projExp.filter((i) => i !== 'state');
    filteredProjExp.push('#A');
    params.ProjectionExpression = filteredProjExp.join(',');
  }
  console.log('fetching agent from dynamo:', params);
  const { Item } = await db.getItem(params).promise();
  if (!Item) {
    return Item;
  }
  console.log('fetched item:', Item);
  return Dynamo.normalize(Item);
};

export const setState = async ({ agentId, agentState, ...attrs }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const additionalAttrs = {};
  const fetchedAgent = await get({ agentId });

  const deletedAttrs = [];
  const neededAttrs = [AGENT_ATTRS.CONNECTION];

  const [stateOnline, stateType, subState] = getStateDef(agentState);
  switch (subState) {
    case AGENT_STATES.DISCONNECTED:
    case AGENT_STATES.OFFLINE:
    case AGENT_STATES.ROUTABLE:
      if (fetchedAgent && fetchedAgent.current_contact_id) {
        console.log(
          `releasing contact id ${fetchedAgent.current_contact_id} from ${agentId}`,
        );
        deletedAttrs.push(AGENT_ATTRS.CURRENT_CONTACT);
        deletedAttrs.push(AGENT_ATTRS.STATE_TTL);
        deletedAttrs.push(AGENT_STATES.CASES);
      }
      break;
    case AGENT_STATES.AGENT_CALLING:
    case AGENT_STATES.AGENT_PENDING:
    case AGENT_STATES.PENDING_CALL:
      console.log('agent requires contact attribute! fetching!');
      neededAttrs.push(AGENT_ATTRS.CURRENT_CONTACT);
      neededAttrs.push(AGENT_ATTRS.STATE_TTL);
      neededAttrs.push(AGENT_ATTRS.CASES);
      break;
    default:
      break;
  }

  if (stateOnline) {
    additionalAttrs[AGENT_ATTRS.ACTIVE] = {
      S: 'y',
    };
  } else {
    deletedAttrs.push(AGENT_ATTRS.ACTIVE);
  }

  console.log('originally passed additional attrs:', attrs);
  const reqKeys = Object.keys(attrs);

  const finalKeys = [...new Set([...reqKeys, ...neededAttrs])].filter(
    (k) => !deletedAttrs.includes(k),
  );

  console.log('AGENT FINAL ATTRS:', finalKeys);

  const getDataType = (attr) => {
    const types = {
      number: 'N',
      string: 'S',
    };
    if (Object.keys(types).includes(typeof attr)) {
      return types[typeof attr];
    }
    return 'S';
  };

  finalKeys.forEach((key) => {
    if (attrs[key] && attrs[key] !== null) {
      additionalAttrs[key] = {
        [getDataType(attrs[key])]: attrs[key],
      };
    } else if (fetchedAgent && fetchedAgent[key]) {
      additionalAttrs[key] = {
        [getDataType(attrs[key])]: fetchedAgent[key],
      };
    }
  });
  const finalAgentState = `${stateOnline}#${stateType}#${subState}`;
  const params = {
    ...KeyMap({
      mapName: 'Item',
      agentId,
      agentState: finalAgentState,
      attributes: {
        state: {
          S: finalAgentState,
        },
        entered_timestamp: {
          S: new Date().toISOString(),
        },
        ...additionalAttrs,
      },
    }),
  };

  console.log('setting state params:', params);
  const results = await db.putItem(params).promise();
  console.log('set agent state: ', agentId, finalAgentState, results);
  const agent = Dynamo.normalize(results[0]);
  console.log('resulting set agent:', agent);

  return {
    namespace: 'phone',
    action: {
      type: 'action',
      name: 'setAgentState',
      data: {
        state: subState,
      },
    },
  };
};

export const getTargetAgent = async ({ currentContactId }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const params = {
    IndexName: 'contact-index',
    ExpressionAttributeValues: {
      ':contactId': {
        S: currentContactId,
      },
    },
    KeyConditionExpression: 'current_contact_id = :contactId',
    Select: 'ALL_ATTRIBUTES',
  };
  console.log('checking for target agent...', params);
  const { Items } = await db.query(params).promise();
  if (!Items.length) {
    console.log('no current target agent!');
    return null;
  }
  const agent = Dynamo.normalize(Items[0]);
  console.log('got target agent!', agent);
  return agent;
};

export class AgentError extends Error {}

export const findNextAgent = async () => {
  const db = Dynamo.DynamoTable(TABLE);
  const params = {
    ExpressionAttributeNames: {
      '#S': 'state',
      '#O': 'active',
    },
    ExpressionAttributeValues: {
      ':t': {
        S: 'online#',
      },
      ':a': {
        S: 'y',
      },
    },
    KeyConditionExpression: '#O = :a AND begins_with (#S, :t)',
    IndexName: 'state-index',
    Select: 'ALL_ATTRIBUTES',
  };
  console.log('querying agents table with:', params);
  const { Items } = await db.query(params).promise();
  const normalizedItems = Items.map((m) => Dynamo.normalize(m));
  console.log('resulting items:', normalizedItems);
  if (!normalizedItems.length) {
    throw new AgentError('No agents are available!');
  }

  const routables = normalizedItems.filter((a) => isRoutable(a.state));
  console.log('filtered routables:');
  if (!routables.length) {
    console.log('active agents found, but none are currently routable!');
    return false;
  }

  // find the agent who's been routable the longest
  const agent = normalizedItems.reduce((pre, cur) => {
    return Date.parse(pre.entered_timestamp) > Date.parse(cur.entered_timestamp)
      ? cur
      : pre;
  });
  console.log('found longest standing routable agent:', agent);
  return agent;
};

export const createStateWSPayload = async ({ agentId, agentState }) => {
  const agent = await get({ agentId });
  return {
    namespace: 'phone',
    action: {
      type: 'action',
      name: 'setAgentState',
    },
    meta: {
      endpoint: process.env.WS_CALLBACK_URL,
      connectionId: agent.connection_id,
    },
    data: {
      state: getStateDef(agentState)[2] || getStateDef(agent.state)[2],
    },
  };
};
