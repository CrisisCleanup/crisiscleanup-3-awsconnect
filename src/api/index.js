/**
 * index.js
 * Api Module
 */

import * as Helpers from './helpers';
import * as Outbound from './outbound';

export { Helpers, Outbound };
