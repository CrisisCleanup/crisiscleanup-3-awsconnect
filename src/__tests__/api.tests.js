/**
 * api.tests.js
 * Api Module Tests
 */

import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { advanceTo, clear } from 'jest-date-mock';
import { Agent, Contact, Events, Helpers, Outbound } from '../api';
import { LANGUAGE } from '../api/helpers';

// jest.mock('../utils/dynamo.js');
jest.mock('../ws');

const MockOutbound = ({ id, phoneNumber, pda, worksite } = {}) => ({
  id: id || 0,
  phone_number: phoneNumber || '+1234567890',
  pda: pda || null,
  worksite: worksite || null,
  external_id: '123abc',
  external_platform: 'connect',
});

const getAxios = () => {
  const mock = new MockAdapter(axios);
  // mock.onHead('http://localhost:8000/shell').reply(200);
  return mock;
};

describe('outbound api', () => {
  it('retrieves outbound by number', async () => {
    const mock = getAxios();
    mock
      .onGet('/phone_outbound', {
        params: { phone_number: '1234567890' },
      })
      .reply(200, {
        results: [MockOutbound()],
      });

    const resp = await Outbound.getByPhoneNumber('+1234567890');
    expect(resp).toMatchSnapshot();
  });

  it('throws error on no results', async () => {
    const mock = getAxios();
    mock
      .onGet('/phone_outbound', {
        params: { phone_number: '1234567890' },
      })
      .reply(404, {
        results: [],
      });
    await expect(Outbound.getByPhoneNumber('123')).rejects.toThrow();
  });

  it('resolves cases by number', async () => {
    const mock = getAxios();
    mock
      .onGet('/phone_outbound', { params: { phone_number: '1234567890' } })
      .reply(200, {
        results: [
          MockOutbound(),
          MockOutbound({ id: 1, worksite: 1 }),
          MockOutbound({ id: 3, pda: 1 }),
        ],
      });
    mock
      .onPost('/phone_connect/resolve_cases', {
        contact_id: 'abc123',
        phone_number: '1234567890',
      })
      .reply(200, {
        results: [{ id: 99 }],
      });
    const cases = await Outbound.resolveCasesByNumber('abc123', '+1234567890');
    expect(cases).toMatchSnapshot();
  });

  it('creates callback', async () => {
    const mock = getAxios();
    mock
      .onPost('/phone_outbound', {
        dnis1: '+10000000000',
        call_type: 'callback',
        language: 2,
        incident_id: ['199'],
        external_id: '123abc',
        external_platform: 'connect',
      })
      .reply(201);
    const resp = await Outbound.create(
      '+10000000000',
      'en_US',
      '199',
      '123abc',
    );
    expect(resp).toMatchInlineSnapshot(`
      Object {
        "config": Object {
          "data": "{\\"dnis1\\":\\"+10000000000\\",\\"call_type\\":\\"callback\\",\\"language\\":2,\\"incident_id\\":[\\"199\\"],\\"external_id\\":\\"123abc\\",\\"external_platform\\":\\"connect\\"}",
          "headers": Object {
            "Accept": "application/json, text/plain, */*",
            "Content-Type": "application/json;charset=utf-8",
          },
          "maxBodyLength": -1,
          "maxContentLength": -1,
          "method": "post",
          "timeout": 0,
          "transformRequest": Array [
            [Function],
          ],
          "transformResponse": Array [
            [Function],
          ],
          "url": "/phone_outbound",
          "validateStatus": [Function],
          "xsrfCookieName": "XSRF-TOKEN",
          "xsrfHeaderName": "X-XSRF-TOKEN",
        },
        "data": undefined,
        "headers": undefined,
        "request": Object {
          "responseUrl": "/phone_outbound",
        },
        "status": 201,
      }
    `);
  });

  it('unlocks the latest callback', async () => {
    const mock = getAxios();
    mock
      .onGet('/phone_outbound', { params: { phone_number: '10001112222' } })
      .reply(200, {
        results: [MockOutbound(), MockOutbound({ id: 99 })],
      });
    mock.onPost('/phone_outbound/99/unlock').reply(200);
    const resp = await Outbound.unlock('+10001112222');
    expect(resp).toMatchSnapshot();
  });

  it('gets the language id', async () => {
    const result = await Helpers.getLanguageId('en_US');
    expect(result).toBe(2);
    const defaultResult = await Helpers.getLanguageId('abc');
    expect(defaultResult).toBe(2);
  });
});

describe('agent api', () => {
  it('generates valid keymaps', () => {
    const keyMap = Agent.KeyMap({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
    });
    expect(keyMap).toMatchSnapshot();
    const itemKeyMap = Agent.KeyMap({
      mapName: 'Item',
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ON_CALL,
      attributes: {
        contactId: 'yyyy',
      },
    });
    expect(itemKeyMap).toMatchSnapshot();
  });

  it('sets agent state', async () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      connection_id: 'zzzz',
    });
    const dbItem = await Agent.get({ agentId: 'xxxx' });
    expect(dbItem).toMatchInlineSnapshot(`
      Object {
        "active": "y",
        "connection_id": "zzzz",
        "entered_timestamp": "2020-06-20T05:00:00.000Z",
        "locale": "en-US",
        "state": "online#routable#routable",
      }
    `);
    advanceTo(new Date(2020, 5, 20, 1, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.OFFLINE,
    });
    const offlineItem = await Agent.get({ agentId: 'xxxx' });
    expect(offlineItem).toMatchInlineSnapshot(`
      Object {
        "active": "y",
        "connection_id": "zzzz",
        "entered_timestamp": "2020-06-20T06:00:00.000Z",
        "locale": "en-US",
        "state": "offline#not_routable#offline",
      }
    `);
    clear();
  });

  it('finds next agent for contact', async () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
    });
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.OFFLINE,
    });
    const nextAgent = await Agent.findNextAgent(LANGUAGE.en_US);
    expect(nextAgent.agent_id).toBe('xxxx');
    clear();
  });

  it('handles online but all non-routable agents', async () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.PAUSED,
    });
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.PAUSED,
    });
    const nextAgent = await Agent.findNextAgent(LANGUAGE.en_US);
    expect(nextAgent).toBeFalsy();
    clear();
  });

  it('handles no online agents', async () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.OFFLINE,
    });
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.OFFLINE,
    });
    await expect(Agent.findNextAgent(LANGUAGE.en_US)).rejects.toThrow(
      Agent.AgentError,
    );
    clear();
  });

  it('correctly finds the longest routable agent', async () => {
    advanceTo(new Date(2019, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
    });
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.ROUTABLE,
    });
    const nextAgent = await Agent.findNextAgent(LANGUAGE.en_US);
    expect(nextAgent.agent_id).toBe('xxxx');
    clear();
  });

  it('correctly filters next agent by language', async () => {
    advanceTo(new Date(2019, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      locale: LANGUAGE.en_US,
    });
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      locale: LANGUAGE.es_MX,
    });
    const nextAgent = await Agent.findNextAgent(LANGUAGE.es_MX);
    expect(nextAgent.agent_id).toBe('yyyy');
    clear();
  });

  it('correctly filters next agent by language bilingual', async () => {
    advanceTo(new Date(2019, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      locale: `${LANGUAGE.en_US}#${LANGUAGE.es_MX}`,
    });
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      locale: LANGUAGE.es_MX,
    });
    const nextAgent = await Agent.findNextAgent(LANGUAGE.es_MX);
    expect(nextAgent.agent_id).toBe('xxxx');
    clear();
  });

  it('correctly finds no agent if none are online with contact locale', async () => {
    advanceTo(new Date(2019, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
      locale: LANGUAGE.en_US,
    });
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'yyyy',
      agentState: Agent.AGENT_STATES.OFFLINE,
      locale: LANGUAGE.es_MX,
    });
    await expect(Agent.findNextAgent(LANGUAGE.es_MX)).rejects.toThrow(
      Agent.AgentError,
    );
    clear();
  });

  it('sustains contact id', async () => {
    advanceTo(new Date(2019, 5, 20, 0, 0, 0, 0));
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.ROUTABLE,
    });
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.PENDING_CALL,
      current_contact_id: 'abc',
    });
    let agentItem = await Agent.get({ agentId: 'xxxx' });
    expect(agentItem).toMatchInlineSnapshot(`
      Object {
        "active": "y",
        "connection_id": "zzzz",
        "current_contact_id": "abc",
        "entered_timestamp": "2019-06-20T05:00:00.000Z",
        "locale": "en-US",
        "state": "online#not_routable#PendingBusy",
      }
    `);
    await Agent.setState({
      agentId: 'xxxx',
      agentState: Agent.AGENT_STATES.BUSY,
    });
    agentItem = await Agent.get({ agentId: 'xxxx' });
    expect(agentItem).toMatchInlineSnapshot(`
      Object {
        "active": "y",
        "connection_id": "zzzz",
        "current_contact_id": "abc",
        "entered_timestamp": "2019-06-20T05:00:00.000Z",
        "locale": "en-US",
        "state": "online#not_routable#Busy",
      }
    `);
    clear();
  });

  it('gets the correct state def', () => {
    const routeState = Agent.getStateDef(Agent.AGENT_STATES.ROUTABLE);
    expect(routeState).toStrictEqual(['online', 'routable', 'routable']);

    expect(Agent.getStateDef('offline')).toStrictEqual([
      'offline',
      'not_routable',
      'offline',
    ]);
    expect(Agent.getStateDef(Agent.AGENT_STATES.NOT_ROUTABLE)).toStrictEqual([
      'online',
      'not_routable',
      'not_routable',
    ]);
    expect(Agent.getStateDef(Agent.AGENT_STATES.PAUSED)).toStrictEqual([
      'online',
      'not_routable',
      'AfterCallWork',
    ]);
    expect(Agent.getStateDef(Agent.AGENT_STATES.AGENT_CALLING)).toStrictEqual([
      'online',
      'not_routable',
      'CallingCustomer',
    ]);
    expect(Agent.getStateDef(undefined)).toStrictEqual([
      'offline',
      'not_routable',
      'offline',
    ]);
  });
});

describe('contact api', () => {
  it('sets correct contact state', async () => {
    const contact = new Contact.Contact({ contactId: 'xxxx' });
    expect(contact.state).toBe('en_US#queued');
    expect(contact.routed).toBe(false);
    await contact.setState('en_US#routed');
    expect(contact.state).toBe('en_US#routed');
    expect(contact.routed).toBe(true);
    await contact.setState('queued');
    expect(contact.state).toBe('en_US#queued');
    await contact.setState('es_MX#routed');
    expect(contact.state).toBe('es_MX#routed');
  });

  it('generates the correct operations', () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    const updateOp = Contact.OPERATIONS.updateContact({
      contactId: 'xxxx',
      state: 'somestate',
      priority: 1,
      cases: {},
    });
    const queryNumOp = Contact.OPERATIONS.queryNumByState({ state: 'routed' });

    expect(updateOp).toMatchSnapshot();
    expect(queryNumOp).toMatchSnapshot();
    clear();
  });

  it('adjusts state based on action', async () => {
    advanceTo(new Date(2020, 5, 20, 0, 0, 0, 0));
    let contact = await new Contact.Contact({ contactId: 'abc' }).load();
    await contact.delete();
    contact = await new Contact.Contact({ contactId: 'abc' }).load();
    contact.action = Contact.CONTACT_ACTIONS.ENTER;
    contact.routed = false;
    await contact.setState(Contact.CONTACT_STATES.QUEUED);
    expect(contact).toMatchSnapshot({ db: expect.anything() });
    contact.action = Contact.CONTACT_ACTIONS.CONNECTING;
    expect(contact.routed).toBe(true);
    await contact.setState();
    expect(contact.routed).toBe(true);
    expect(contact).toMatchSnapshot({ db: expect.anything() });
    expect(contact.state).toBe('en_US#routed');
    clear();
  });
});

describe('events api', () => {
  it('creates expected event key', () => {
    const event = new Events.Event({ itemId: 'xxxx' });
    event.object(Events.EVENT_OBJECTS.AGENT).update();
    expect(event.eventKey).toBe('update_agent');
    const callEvent = new Events.Event({ itemId: 'call_xxxx' })
      .object(Events.EVENT_OBJECTS.INBOUND)
      .join(event);
    expect(callEvent.eventKey).toBe('join_inbound-call_to_agent');
  });
});
