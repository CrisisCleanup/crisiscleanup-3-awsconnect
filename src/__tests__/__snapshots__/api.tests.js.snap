// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent api generates valid keymaps 1`] = `
Object {
  "Key": Object {
    "agent_id": Object {
      "S": "xxxx",
    },
  },
}
`;

exports[`agent api generates valid keymaps 2`] = `
Object {
  "Item": Object {
    "agent_id": Object {
      "S": "xxxx",
    },
    "contactId": "yyyy",
  },
}
`;

exports[`agent api sets agent state 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "hash": "agent_id",
        "name": "connect-agents-table-dev",
      },
    ],
    Array [
      Object {
        "hash": "agent_id",
        "name": "connect-agents-table-dev",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "putItem": [MockFunction],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "putItem": [MockFunction],
      },
    },
  ],
}
`;

exports[`agent api sets agent state 2`] = `
Object {
  "calls": Array [],
  "instances": Array [],
  "invocationCallOrder": Array [],
  "results": Array [],
}
`;

exports[`contact api generates the correct operations 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#E": "ttl",
    "#P": "priority",
    "#S": "state",
    "#T": "entered_timestamp",
  },
  "ExpressionAttributeValues": Object {
    ":e": 1592629800,
    ":p": 1,
    ":s": "somestate",
    ":t": "2020-06-20T05:00:00.000Z",
  },
  "Key": Object {
    "contact_id": "xxxx",
  },
  "UpdateExpression": "set #S = :s, #P = :p, #T = :t, #E = :e",
}
`;

exports[`outbound api creates callback 1`] = `
Object {
  "config": Object {
    "data": "{\\"dnis1\\":\\"+10000000000\\",\\"call_type\\":\\"callback\\",\\"language\\":5,\\"incident_id\\":[\\"199\\"],\\"external_id\\":\\"123abc\\",\\"external_platform\\":\\"connect\\"}",
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json;charset=utf-8",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 0,
    "transformRequest": Array [
      [Function],
    ],
    "transformResponse": Array [
      [Function],
    ],
    "url": "/phone_outbound",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": undefined,
  "headers": undefined,
  "request": Object {
    "responseUrl": "/phone_outbound",
  },
  "status": 201,
}
`;

exports[`outbound api resolves cases by number 1`] = `
Object {
  "ids": Array [
    0,
    1,
    3,
  ],
  "pdas": Array [
    1,
  ],
  "worksites": Array [
    1,
  ],
}
`;

exports[`outbound api retrieves outbound by number 1`] = `
Array [
  Object {
    "external_id": "123abc",
    "external_platform": "connect",
    "id": 0,
    "pda": null,
    "phone_number": "+1234567890",
    "worksite": null,
  },
]
`;

exports[`outbound api unlocks the latest callback 1`] = `
Object {
  "config": Object {
    "data": undefined,
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 0,
    "transformRequest": Array [
      [Function],
    ],
    "transformResponse": Array [
      [Function],
    ],
    "url": "/phone_outbound/99/unlock",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": undefined,
  "headers": undefined,
  "request": Object {
    "responseUrl": "/phone_outbound/99/unlock",
  },
  "status": 200,
}
`;
